types/wlr_pointer_constraints_v1.c

```diff
@@ -238,38 +263,93 @@ static void pointer_constraints_bind(struct wl_client *client, void *data,
 	assert(client && cons);
 
 	struct wl_resource *resource = wl_resource_create(client,
-		&zwp_pointer_constraints_v1_interface, version, id);
+			&zwp_pointer_constraints_v1_interface, version, id);
 	if (resource == NULL) {
 		wl_client_post_no_memory(client);
 		return;
 	}
 
 	wl_resource_set_implementation(resource, &pointer_constraints_impl,
-		cons, NULL);
+			cons, NULL);
 }
 
-struct wlr_pointer_constraints_v1 *wlr_pointer_constraints_v1_create(struct wl_display *display) {
-	struct wlr_pointer_constraints_v1 *cons =
-		calloc(1, sizeof(struct wlr_pointer_constraints_v1));
-	if (!cons) {
-		return NULL;
-	}
-
+bool wlr_pointer_constraints_v1_init(struct wlr_pointer_constraints_v1 *cons,
+		const struct wlr_pointer_constraints_v1_interface *impl, struct wl_display *display) {
 	struct wl_global *wl_global = wl_global_create(display,
 		&zwp_pointer_constraints_v1_interface, 1, cons, pointer_constraints_bind);
 	if (!wl_global) {
 		free(cons);
-		return NULL;
+		return false;
 	}
 	cons->wl_global = wl_global;
+	cons->impl = impl;
 
-	wl_signal_init(&cons->events.constrain);
+	wlr_log(L_DEBUG, "wlr_pointer_constraints_v1_init(%p, %p, %p) with global %p", cons, impl, display, wl_global);
 
-	return cons;
+	return true;
 }
 
-void wlr_pointer_constraints_v1_destroy(struct wlr_pointer_constraints_v1 *cons) {
-	if(!cons) return;
+void wlr_pointer_constraints_v1_finish(struct wlr_pointer_constraints_v1 *cons) {
 	wl_global_destroy(cons->wl_global);
-	free(cons);
+}
+
+// checks whether we are within the region
+void wlr_pointer_constraints_v1_check_constraint_active(
+		struct wlr_pointer_constraints_v1 *pointer_constraints,
+		struct wlr_seat *seat, double sx, double sy) {
+	struct wlr_pointer_constraint_v1 *constraint = seat->current_pointer_constraint;
+
+	// any of these things being true means
+	// we don't have to check whether we
+	// have entered the region of the pointer constraint
+	if (
+			!constraint || // no constraint to check for
+			!constraint->accepted || // even if we are in it we can't activate it
+			!constraint->current.region || // it's always active if accepted
+			constraint->active) { // if it's already active it can't be deactived here
```

Comment:
Instead of doing this inline with this crazy if statement, just summarize the criteria in a standalone comment above the condition. Use `/* */` for multi-line comments.